### ISSUE TITLE
Logging does not work when running multiple trainings on a single node in parallel

### ISSUE THREAD

**Issue by AreopagX** [Lightning-AI/pytorch-lightning#19318]
### Bug description

I'm working on a SLURM cluster where each node has 8 gpus. What I wanna do is to train 8 models in parallel, each on it's own gpu, independent of each other. When I start the training, the folder containing the lightning logs is created. Within the respective folders, the config.yaml and the checkpoints are saved. However, only the folder from task 0 contains a events.out file from the TensorBoard Logger as well as the hparams.yaml. I'd expect that every directory contains a log-file for the respective task as well.

### What version are you seeing the problem on?

v2.1

### How to reproduce the bug

Running the following script using srun on the SLURM cluster. The script is run on 1 node, 8 tasks are created with 1 gpu per task and 8 cpu cores per task. 

``` python
    slurm_job_id = os.environ.get("SLURM_JOB_ID", "0")
    slurm_id = os.environ.get("SLURM_LOCALID", 0)
    cli = LightningCLI(
        args=[
            "fit",
            "--config", config_file,  # contains "trainer.logger: null"
            "--trainer.strategy", "lightning.pytorch.strategies.SingleDeviceStrategy",
            "--trainer.strategy.init_args.device", f"cuda:{slurm_id}",
            "--trainer.logger", "lightning.pytorch.loggers.tensorboard.TensorBoardLogger",
            "--trainer.logger.init_args.save_dir", ".",
            "--trainer.logger.init_args.name", f"lightning_logs",
            "--trainer.logger.init_args.version", f"job_{slurm_job_id}",
            "--trainer.logger.init_args.sub_dir", f"slurm_{slurm_id}"
        ]
    )
```

### Error messages and logs

N/A

### Environment

#- Lightning Component (Trainer, TensorBoardLogger, LightningCLI):
#- PyTorch Lightning Version (e.g., 1.5.0): 2.1.0
#- PyTorch Version (e.g., 2.0): 2.1.0+rocm5.6
#- Python version (e.g., 3.9): 3.10
#- OS (e.g., Linux): Linux
#- CUDA/cuDNN version: ROCm 5.6
#- GPU models and configuration: 8x AMD Instinct MI210
#- How you installed Lightning(`conda`, `pip`, source): pip
#- Running environment of LightningApp (e.g. local, cloud): SLURM node

### More info

When debugging this issue I found out that for all tasks but task 0, the LightningModule.logger variable looked as follows:

``` 
{'_root_dir': '.', '_name': 'lightning_logs', '_version': 'job_1182700', '_sub_dir': 'slurm_7', '_default_hp_metric': True, '_prefix': '', '_fs': <fsspec.implementations.local.LocalFileSystem object at 0x153d6b56d5a0>, '_experiment': None, '_kwargs': {'comment': '', 'purge_step': None, 'max_queue': 10, 'flush_secs': 120, 'filename_suffix': ''}, '_log_graph': False, 'hparams': {}}
```

While in task 0 it looks like:

```
{'_root_dir': '.', '_name': 'lightning_logs', '_version': 'job_1182700', '_sub_dir': 'slurm_0', '_default_hp_metric': True, '_prefix': '', '_fs': <fsspec.implementations.local.LocalFileSystem object at 0x1470fba0b2b0>, '_experiment': <torch.utils.tensorboard.writer.SummaryWriter object at 0x1470fbd25a80>, '_kwargs': {'comment': '', 'purge_step': None, 'max_queue': 10, 'flush_secs': 120, 'filename_suffix': ''}, '_log_graph': False, 'hparams': {}}
```

Why is the Summary Writer none for tasks 1 to 7?

```
$ ls -R lightning_logs/job_1182700/

lightning_logs/job_1182700/:
slurm_0  slurm_1  slurm_2  slurm_3  slurm_4  slurm_5  slurm_6  slurm_7

lightning_logs/job_1182700/slurm_0:
config.yaml  events.out.tfevents.1705702465.node.30161.0  hparams.yaml

lightning_logs/job_1182700/slurm_1:
config.yaml

lightning_logs/job_1182700/slurm_2:
config.yaml

lightning_logs/job_1182700/slurm_3:
config.yaml

lightning_logs/job_1182700/slurm_4:
config.yaml

lightning_logs/job_1182700/slurm_5:
config.yaml

lightning_logs/job_1182700/slurm_6:
config.yaml

lightning_logs/job_1182700/slurm_7:
config.yaml
```

cc @awaelchli

**Comment by AreopagX** [comment]
Doing some more digging I find that in [Lightning Fabric Tensorboard](https://github.com/Lightning-AI/pytorch-lightning/blob/e89f46a74e9541ba832c3ade1eb3f87ef072e844/src/lightning/fabric/loggers/tensorboard.py#L176C1-L197C32) a rank zero check prevents that logger._experiment is set when the current rank is not zero. However, I don't know why this is not the case. Since I'm using the Single Device Strategy, it should always be on global zero, shouldn't it?

When I overwrite the experiment() property in the [TensorBoardLogger](https://github.com/Lightning-AI/pytorch-lightning/blob/e89f46a74e9541ba832c3ade1eb3f87ef072e844/src/lightning/pytorch/loggers/tensorboard.py#L42) class the experiment files are created. Yet, the training runs do not log their data to this logger. I guess this is also due to rank zero checks, but I don't know where these happen.

**Comment by awaelchli** [comment]
@AreopagX Can you share the submission script please? Make sure that you follow [this guide](https://lightning.ai/docs/pytorch/stable/clouds/cluster_advanced.html) **exactly**.  If you want to run independent jobs, then certainly the job id should be different for each, and that's not what we see in your logs. So I think you didn't launch the job correctly. You probably accidentally launched one job with a world size of 8, instead of 8 jobs with world size 1 each.

**Comment by AreopagX** [comment]
Thanks for your response @awaelchli . Here is the submission script:

```bash
#!/bin/bash -l
#SBATCH --job-name Job
#SBATCH --output=%j.out
#SBATCH --error=%j.err
#SBATCH --partition=gpu
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=8
#SBATCH --cpus-per-task=8
#SBATCH --mem=0
#SBATCH --no-requeue
#SBATCH --time=10:00:00
#SBATCH --chdir=/my/directory

conda activate /my/directory/env

srun python slurm.py
```

It is intended that I'm running a single job only, since AFAIK I cannot submit multiple of jobs to be run on the same node simultaneously. Since the tasks only require 1 gpu per task, I'd be having 7 gpus idle while one training is blocking the node. So my idea was to use multiple tasks on a single node, so that each gpu has its own task and training associated with it. Please tell me if I'm wrong, I'm only using SLURM for a week or so.

**Comment by awaelchli** [comment]
Ok I understand. Yes this is not really supported, because `--ntasks-per-node=8` will normally be interpreted as the number of processes that SLURM launches for multi-gpu training. What you want can probably achieved with a hack by disabling the Slurm detection in Lightning like so:

```py
from lightning.pytorch.plugins import LightningEnvironment

trainer = Trainer(
    devices=x,  # probably need to set x = slurm_task_id here I don't know
    plugins=LightningEnvironment(),  # force non-slurm environment
)
```

**Comment by AreopagX** [comment]
Thanks for your suggestion, passing `--trainer.plugins.class_path lightning.pytorch.plugins.environments.LightningEnvironment` to the Lightning CLI made it work.
